src/poly/mod.rs

```diff
@@ -332,15 +418,42 @@ pub trait SignalFactory<V> {
 
 /// The result of decomposing a PI into several
 #[derive(Debug, Clone)]
-pub struct ConstrDecomp<F, V> {
+pub struct ConstrDecomp<F, V, M> {
     /// PI constraint for the new signals introduced.
-    pub constrs: Vec<Expr<F, V, ()>>,
+    pub constrs: Vec<Expr<F, V, M>>,
     /// Expressions for how to create the witness for the generated signals the orginal expression
     /// has be decomposed into.
-    pub auto_signals: HashMap<V, Expr<F, V, ()>>,
+    pub auto_signals: HashMap<V, Expr<F, V, M>>,
 }
 
-impl<F, V> Default for ConstrDecomp<F, V> {
+impl<F, V: Debug, M> ConstrDecomp<F, V, M> {
+    pub fn get_auto_signal<S: Into<String> + Copy>(
+        &self,
+        annotation: S,
+    ) -> Option<(&V, &Expr<F, V, M>)> {
+        self.auto_signals.iter().find_map(|(s, e)| {
+            if format!("{:#?}", s) == annotation.into() {
+                Some((s, e))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl<F: Clone, V: Clone> ConstrDecomp<F, V, HashResult> {
+    pub fn find_auto_signal_by_hash(&self, hash: u64) -> Option<(&V, &Expr<F, V, HashResult>)> {
```

Comment:
It is not being used right now, but I think it can be useful anytime, and this was to look into the `ConstrDecomp` and find if there was already an `auto_signal` created for that expression, because this was the most certain way to know if there was already a signal for that expression or not